File: rust/2015/day08a/src/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn process_file(filename: &str) -> io::Result<u32> {
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    let mut result: u32 = 0;
    for line_result in reader.lines() {
        let line = line_result?;
        let code_len = line.len() as u32;
        let mut mem_len = 0;
        let mut i: usize = 1;
        while i < line.len() - 1 {
            match line.chars().nth(i).unwrap() {
                '\\' => {
                    match line.chars().nth(i + 1).unwrap() {
                        '\\' | '\"' => i += 2,
                        'x' => i += 4,
                        _ => i += 1,
                    }
                }
                _ => i += 1,
            }
            mem_len += 1;
        }
        result += code_len - mem_len;
    }
    Ok(result)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}

File: zig/2015/day08a/src/main.zig
const std = @import("std");

fn usage(progname: []const u8) void {
    std.debug.print("usage: {s} <input file>\n", .{progname});
    std.process.exit(1);
}

fn processFile(filename: []const u8) !u32 {
    var file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    var read_buf: [131072]u8 = undefined;
    var reader = file.reader(&read_buf);
    var r_interface = &reader.interface;

    var result: u32 = 0;

    while ( true ) {
        const line = r_interface.takeDelimiterExclusive('\n') catch |err| switch(err) {
            error.EndOfStream => break,
            else => return err,
        };
        const code_len: u32 = @intCast(line.len);
        var mem_len: u32 = 0;
        var i: usize = 1;
        while ( i < line.len - 1 ) {
            switch ( line[i] ) {
                '\\' =>
                    switch ( line[i + 1] ) {
                        '\\', '\"' => i += 2,
                        'x' => i += 4,
                        else => i += 1,
                    },
                else => i += 1,
            }
            mem_len += 1;
        }
        result += code_len - mem_len;
    }

    return result;
}

pub fn main() !void {
    var stdout_buffer: [1024]u8 = undefined;
    const allocator = std.heap.page_allocator;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buffer);
    var stdout = &stdout_writer.interface;
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if ( args.len < 2 ) {
        usage(args[0]);
    }

    const filename = args[1];

    const result = processFile(filename) catch |err| {
        std.debug.print("error while processing file `{s}': {s}\n", .{filename, @errorName(err)});
        std.process.exit(1);
    };

    try stdout.print("result = {}\n", .{result});
    try stdout.flush();
}
